File: Sorting/MergeSort.java
import java.util.*;
public class MergeSort extends SortParent {

    private int length;
    private int[] arr;
    private int[] tempArr;

    public static void main(String[] args) {
        MergeSort ms = new MergeSort();
        ms.sort();
    }

    /**
    */
    public int[] sort() {
        this.length = numbers.length;
        this.arr = numbers;
        int[] tempArr = new int[length];
        mergeSort(0, length -1);
        return numbers;
    }

    private void mergeSort(int lowerIndex, int higherIndex) {
        if (lowerIndex < higherIndex) {
            int middle = lowerIndex + (higherIndex - lowerIndex) / 2;
            // do the left side
            mergeSort(lowerIndex, middle);
            // do the right side
            mergeSort(middle + 1, higherIndex);
            // merge those parts together
            mergeParts(lowerIndex, middle, higherIndex);
        }
    }

    private void mergeParts(int lowerIndex, int middle, int higherIndex) {
        for (int i = lowerIndex; i <= higherIndex; i++) {
            tempArr[i] = arr[i];
        }

        int i = lowerIndex;
        int j = middle + 1;
        int k = lowerIndex;
        while (i <= middle && j <= higherIndex) {
            if (tempArr[i] <= tempArr[j]) {
                arr[k] = tempArr[i];
                i++;
            } else {
                arr[k] = tempArr[j];
                j++;
            }
            k++;
        }
        while (i <= middle) {
            arr[k] = tempArr[i];
            k++;
            i++;
        }
    }
}

File: Sorting/MergeSort.py
class MergeSort(object):

    def merge_sort(self, arr):
        if len(arr) > 1:
            mid = len(arr) / 2
            left = arr[:mid]
            right = arr[mid:]

            # recurse
            self.merge_sort(left)
            self.merge_sort(right)

            i = j = k = 0

            while i < len(left) and j < len(right):
                if left[i] < right[j]:
                    arr[k] = left[i]
                    i = i + 1
                else:
                    arr[k] = right[j]
                    j = j + 1
                k = k + 1

            while i < len(left):
                arr[k] = left[i]
                k = k + 1
                i = i + 1

            while j < len(left):
                arr[k] = left[j]
                k = k + 1
                j = j + 1
        print arr



ms = MergeSort()
arr = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
ms.merge_sort(arr)
